File: apps/mobile-app-flutter/lib/upload_flow.dart
import 'dart:typed_data';
import 'package:databiomics_sdk_dart/upload_client.dart';

Future<Map<String, dynamic>> uploadBytesWithTus({
  required UploadApiDart api,
  required String workspaceId,
  required String userId,
  required String uploadType,
  required Uint8List bytes,
  String? existingSessionId,
}) async {
  const chunkSize = 256 * 1024;
  var sessionId = existingSessionId;
  var offset = 0;

  if (sessionId != null) {
    final status = await api.getTusSessionOffset(sessionId);
    if (status['status'] == 404) {
      sessionId = null;
    } else if (status['status'] == 204) {
      final serverOffset = int.tryParse('${status['offset'] ?? 0}');
      final serverLength = int.tryParse('${status['length'] ?? bytes.length}');
      if (serverOffset == null || serverLength == null) {
        throw Exception('Resposta inválida do servidor para offset/length da sessão TUS');
      }
      if (serverLength != bytes.length) {
        throw Exception('Tamanho do arquivo diverge da sessão TUS existente');
      }
      if (serverOffset < 0 || serverOffset > bytes.length) {
        throw Exception('Offset retornado pelo servidor está fora do intervalo esperado');
      }
      offset = serverOffset;
    } else {
      throw Exception('Falha ao consultar sessão TUS existente: HTTP ${status['status']}');
    }
  }

  if (sessionId == null) {
    final created = await api.createTusSession({
      'workspace_id': workspaceId,
      'user_id': userId,
      'upload_type': uploadType,
      'expected_size': bytes.length,
      'metadata': {'filename': 'mobile-upload.bin'}
    });
    sessionId = created['session_id'] as String;
  }

  while (offset < bytes.length) {
    final end = (offset + chunkSize > bytes.length) ? bytes.length : offset + chunkSize;
    final chunk = Uint8List.sublistView(bytes, offset, end);
    final status = await api.uploadChunk(sessionId!, offset, chunk);
    if (status != 204) {
      throw Exception('Falha ao enviar chunk no offset $offset');
    }
    offset = end;
  }

  return api.complete(sessionId!);
}

File: apps/frontend-web/src/upload-flow.ts
import { UploadApiTs } from '../../../packages/sdk-ts/src/upload-client.js';

export async function uploadFileWithTus(params: {
  apiBaseUrl: string;
  token: string;
  workspaceId: string;
  userId: string;
  file: File;
  uploadType: 'avatar' | 'audio' | 'video' | 'deck_asset';
  existingSessionId?: string;
}) {
  const api = new UploadApiTs(params.apiBaseUrl, params.token);
  const chunkSize = 256 * 1024;
  let sessionId = params.existingSessionId;
  let offset = 0;

  if (sessionId) {
    const status = await api.getTusSessionOffset(sessionId);
    if (status.status === 404) {
      sessionId = undefined;
    } else if (status.status === 204) {
      const serverOffset = Number(status.offset ?? 0);
      const serverLength = Number(status.length ?? params.file.size);
      if (Number.isNaN(serverOffset) || Number.isNaN(serverLength)) {
        throw new Error('Resposta inválida do servidor para offset/length da sessão TUS');
      }
      if (serverLength !== params.file.size) {
        throw new Error('Tamanho do arquivo diverge da sessão TUS existente');
      }
      if (serverOffset < 0 || serverOffset > params.file.size) {
        throw new Error('Offset retornado pelo servidor está fora do intervalo esperado');
      }
      offset = serverOffset;
    } else {
      throw new Error(`Falha ao consultar sessão TUS existente: HTTP ${status.status}`);
    }
  }

  if (!sessionId) {
    const created = await api.createTusSession({
      workspace_id: params.workspaceId,
      user_id: params.userId,
      upload_type: params.uploadType,
      expected_size: params.file.size,
      metadata: { filename: params.file.name }
    });

    sessionId = created.session_id as string;
  }

  while (offset < params.file.size) {
    const chunk = params.file.slice(offset, offset + chunkSize);
    const bytes = new Uint8Array(await chunk.arrayBuffer());
    const patch = await api.uploadChunk(sessionId, offset, bytes);
    if (patch.status !== 204) {
      throw new Error(`Falha no upload chunk em ${offset}`);
    }
    const nextOffset = Number(patch.nextOffset);
    offset = Number.isNaN(nextOffset) ? offset + bytes.length : nextOffset;
  }

  return api.complete(sessionId);
}
